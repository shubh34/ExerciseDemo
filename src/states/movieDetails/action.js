import { RSAA } from 'redux-api-middleware';
import endPoints from '../../configs/endPoints';


export const GET_MOVIE_DETAILS_REQUEST = 'GET_MOVIE_DETAILS_REQUEST';
export const GET_MOVIE_DETAILS_SUCCESS = 'GET_MOVIE_DETAILS_SUCCESS';
export const GET_MOVIE_DETAILS_FAILURE = 'GET_MOVIE_DETAILS_FAILURE';

export const RESET_MOVIE_DETAILS = 'RESET_MOVIE_DETAILS';

export const getMovieDetails = (movieName) => {
	const endPoint = endPoints.getMovieDetails;
	const url = endPoint.resolve(movieName);
	return {
		[RSAA]: {
			endpoint: url,
			method: endPoint.method,
			types: [GET_MOVIE_DETAILS_REQUEST, GET_MOVIE_DETAILS_SUCCESS, GET_MOVIE_DETAILS_FAILURE],
		},
	};
};

export const resetMovieDetails = () => ({
	type: RESET_MOVIE_DETAILS,
});
