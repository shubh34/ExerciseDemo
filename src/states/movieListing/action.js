import { RSAA } from 'redux-api-middleware';
import endPoints from '../../configs/endPoints';
import { cacheApiResponse } from '../../utils/cacheApiUtils';

export const GET_MOVIE_LIST_BY_CATEGORY_REQUEST = 'GET_MOVIE_LIST_BY_CATEGORY_REQUEST';
export const GET_MOVIE_LIST_BY_CATEGORY_SUCCESS = 'GET_MOVIE_LIST_BY_CATEGORY_SUCCESS';
export const GET_MOVIE_LIST_BY_CATEGORY_FAILURE = 'GET_MOVIE_LIST_BY_CATEGORY_FAILURE';

export const createRSAAMovieListReuest = (categoryName) => {
	const endPoint = endPoints.fetchMovieListByCategory;
	const url = endPoint.resolve(categoryName);
	const success = { type: GET_MOVIE_LIST_BY_CATEGORY_SUCCESS, meta: { id: categoryName, cacheKey: GET_MOVIE_LIST_BY_CATEGORY_REQUEST } };
	const failure = { type: GET_MOVIE_LIST_BY_CATEGORY_FAILURE, meta: { id: categoryName } };
	return {
		[RSAA]: {
			endpoint: url,
			method: endPoint.method,
			types: [GET_MOVIE_LIST_BY_CATEGORY_REQUEST, success, failure],
		},
	};
};

export const fetchMovieListByCategory = categoryName => dispatch => dispatch(createRSAAMovieListReuest(categoryName)).then((response) => {
	cacheApiResponse(response);
});
