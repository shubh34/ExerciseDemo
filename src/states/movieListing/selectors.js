import get from 'lodash/get';

export const getMovieListHeader = (state, id) => get(state, `movieListing[${id}].data.name`);

export const getMovieListByCategory = (state, id) => get(state, `movieListing[${id}].data.contents.data`, []).map(movie => ({
	artwork: movie.images.artwork.replace('.jpeg', '-width217-quality80.jpeg'),
	id: movie.id,
	score: movie.highlighted_score.score,
	votes: movie.highlighted_score.formatted_amount_of_votes,
}));
