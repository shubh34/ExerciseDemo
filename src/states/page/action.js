
export const GO_TO_PAGE = 'GO_TO_PAGE';


export const goToPage = name => ({
	type: GO_TO_PAGE,
	name,
});
