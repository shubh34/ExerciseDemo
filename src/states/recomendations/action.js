import { RSAA } from 'redux-api-middleware';
import endPoints from '../../configs/endPoints';

export const GET_MOVIE_RECOMENDATION_REQUEST = 'GET_MOVIE_RECOMENDATION_REQUEST';
export const GET_MOVIE_RECOMENDATION_SUCCESS = 'GET_MOVIE_RECOMENDATION_SUCCESS';
export const GET_MOVIE_RECOMENDATION_FAILURE = 'GET_MOVIE_RECOMENDATION_FAILURE';

export const fetchRecomendationForMovie = id => {
	const endPoint = endPoints.getRecomendationForMovie;
	const url = endPoint.resolve(id);
	return {
		[RSAA]: {
			endpoint: url,
			method: endPoint.method,
			headers: {
				Accept: 'application/json',
				'content-type': 'application/json'
			},
			types: [GET_MOVIE_RECOMENDATION_REQUEST, GET_MOVIE_RECOMENDATION_SUCCESS, GET_MOVIE_RECOMENDATION_FAILURE]
		}
	};
};
