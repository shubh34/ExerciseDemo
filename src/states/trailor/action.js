import { RSAA } from 'redux-api-middleware';
import endPoints from '../../configs/endPoints';

export const GET_MOVIE_TRAILOR_REQUEST = 'GET_MOVIE_TRAILOR_REQUEST';
export const GET_MOVIE_TRAILOR_SUCCESS = 'GET_MOVIE_TRAILOR_SUCCESS';
export const GET_MOVIE_TRAILOR_FAILURE = 'GET_MOVIE_TRAILOR_FAILURE';


export const getMovieTrailor = () => {
	const endPoint = endPoints.getMovieTrailer;
	const url = endPoint.resolve();
	return {
		[RSAA]: {
			endpoint: url,
			method: endPoint.method,
			headers: {
				Accept: 'application/json',
				'content-type': 'application/json',

			},
			body: JSON.stringify({
				audio_language: 'SPA',
				audio_quality: '2.0',
				content_id: 'matrix',
				content_type: 'movies',
				device_serial: 'device_serial_1',
				device_stream_video_quality: 'FHD',
				player: 'web:PD-NONE',
				subtitle_language: 'MIS',
				video_type: 'trailer',
			}),
			types: [GET_MOVIE_TRAILOR_REQUEST, GET_MOVIE_TRAILOR_SUCCESS, GET_MOVIE_TRAILOR_FAILURE],
		},
	};
};
