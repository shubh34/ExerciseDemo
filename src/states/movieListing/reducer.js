import { GET_MOVIE_LIST_BY_CATEGORY_SUCCESS } from './action';


const reducer = (state = [], action) => {
	switch (action.type) {
	case GET_MOVIE_LIST_BY_CATEGORY_SUCCESS:
		console.log(action.payload);
	default:
		console.log('state', state);
		return state;
	}
};
export default reducer;
