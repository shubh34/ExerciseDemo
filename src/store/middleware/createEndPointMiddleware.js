/* eslint-disable import/no-mutable-exports */
import { isRSAA, RSAA } from 'redux-api-middleware';
import { GET_MOVIE_LIST_BY_CATEGORY_SUCCESS } from '../../states/movieListing/action';

export const apiBaseUrl = 'https://gizmo.rakuten.tv/v3/';

export const apiToBeCached = [GET_MOVIE_LIST_BY_CATEGORY_SUCCESS];

export const getActionTypesWithCacheKey = (types, endPointWithQueryParam) =>
	types.map(action => {
		if (action instanceof Object && apiToBeCached.includes(action.type)) {
			return {
				...action,
				meta: {
					...action.meta,
					cacheKey: endPointWithQueryParam
				}
			};
		}
		return action;
	});

export const endPointMiddleWare = () => () => next => action => {
	if (!isRSAA(action)) {
		return next(action);
	}
	const endPoint = action[RSAA].endpoint;
	const endPointWithBAseUrl = apiBaseUrl + endPoint;
	const endPointWithQueryParam = `${endPointWithBAseUrl}?classification_id=5&device_identifier=web&locale=es&market_code=es`;
	const types = getActionTypesWithCacheKey(action[RSAA].types, endPointWithQueryParam);
	const nextAction = {
		[RSAA]: {
			...action[RSAA],
			types,
			endpoint: endPointWithQueryParam
		}
	};

	return next(nextAction);
};
