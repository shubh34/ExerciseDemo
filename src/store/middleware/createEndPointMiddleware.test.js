import { RSAA } from 'redux-api-middleware';
import { endPointMiddleWare, apiBaseUrl, getActionTypesWithCacheKey } from './createEndPointMiddleware';
import { GET_MOVIE_LIST_BY_CATEGORY_SUCCESS } from '../../states/movieListing/action';


const setup = (
	next = jest.fn(),
) => ({
	next,
	middleware: endPointMiddleWare()()(next),
});

describe('createEndpointMiddleware', () => {
	it('calls next with original action if not RSAA action', () => {
		const { middleware, next } = setup();
		const action = {
			types: 'DEMO_TEST',
		};

		middleware(action);
		expect(next).toBeCalledWith(action);
	});

	it('should add apiBase Url and query param', () => {
		const { middleware, next } = setup();

		const action = {
			[RSAA]: {
				endpoint: 'test/RSAA',
				types: [],
			},
		};

		middleware(action);

		expect(next).toBeCalledWith(
			expect.objectContaining({
				[RSAA]: {
					endpoint: `${apiBaseUrl}test/RSAA?classification_id=5&device_identifier=web&locale=es&market_code=es`,
					types: [],
				},
			}),
		);
	});
	describe('getActionTypesWithCacheKey', () => {
		it('should not return typeWithCacheKey when api not to be cached', () => {
			expect(getActionTypesWithCacheKey([{ type: 'test' }], 'cacheKey')).toMatchSnapshot();
		});
		it('should return typeWithCacheKey when api is to be cached', () => {
			expect(getActionTypesWithCacheKey([{ type: GET_MOVIE_LIST_BY_CATEGORY_SUCCESS }], 'cacheKey')).toMatchSnapshot();
		});
	});
});
