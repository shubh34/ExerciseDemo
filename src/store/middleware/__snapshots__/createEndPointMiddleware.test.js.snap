// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createEndpointMiddleware getActionTypesWithCacheKey should not return typeWithCacheKey when api not to be cached 1`] = `
Array [
  Object {
    "type": "test",
  },
]
`;

exports[`createEndpointMiddleware getActionTypesWithCacheKey should return typeWithCacheKey when api is to be cached 1`] = `
Array [
  Object {
    "meta": Object {
      "cacheKey": "cacheKey",
    },
    "type": "GET_MOVIE_LIST_BY_CATEGORY_SUCCESS",
  },
]
`;
